[expo-image] Add resizeMode support

IjzerenHein opened the pull request:
# Why

`expo-image` should resize according to the `resizeMode` prop. When no `resizeMode` is provided, the image uses `cover`.

# How

## JS
- `resizeMode` prop is removed from style

## iOS
- Setting `UIImageView.contentMode` adds support for `cover`, `contain` and `stretch`
- Setting the SDWebImage scale-factor was required for `center` and `repeat`
- Adding a custom load completion has was needed to set the "tile" mode on the UIImage, in order to make `repeat` work

## Android
- Setting the `ImageView.scaleType` and providing Glide with the option `fitCenter` adds support for `cover`, `contain` and `stretch`
- Overriding the image-size in the Glide options add support for `center`
- `repeat` is not yet supported

## Web
- No effort was needed

# Test Plan

- All images (also Animated gifs) now render with the same size as `<Image>` (tested with the test-app in NCL)
- Overflow issues are out of scope of this PR
- The only exception is `repeat` on Android with requires additional effort and is not supported yet

![Simulator Screen Shot - iPhone 11 Pro Max - 2020-03-16 at 10 41 06](https://user-images.githubusercontent.com/6184593/76743285-afb48f00-6772-11ea-957e-2dead5c4ae82.png)
![Screenshot_1584352026](https://user-images.githubusercontent.com/6184593/76743833-86483300-6773-11ea-9208-67b04ebacb52.png)

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
-  [self sd_setImageWithURL:imageURL];
+- (void)setResizeMode:(RCTResizeMode)resizeMode
+{
+  if (_resizeMode == resizeMode) return;
```
Let's write out full-sized `if` with brackets (💅)

IjzerenHein replied:
You're right, will do 👍

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
+{
+  _needsReload = NO;
+  
+  NSURL *imageUrl = _source ? [RCTConvert NSURL:_source[sourceUriKey]] : nil;
```
At first I also used `imageUrl`, but then I noticed that both `NSURL` treats `URL` as full uppercase and there is a `sd_imageURL`, so I figured that must be the Obj-C way of naming URLs (💅)

IjzerenHein replied:
Ah okay. Well this PR focused on the resizeMode part, but I had to make some changes to the loading as well to make that work. We should probably use your loading code from the events PR, and then I can integrate the resizeMode changes in there.

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
+        image = [image resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];
+      }
+    }
+    self.image = image;
```
Strongly catching `self` may leak its instance. Somehow we guard against that in `dealloc`, cancelling all operations, but still, we should use _weak self_ pattern.

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
+    
+    // Modifications to the image like changing the resizing-mode or cap-insets
+    // cannot be handled using a SDWebImage transformer, because they don't change
+    // the image-data and this causes this "meta" data to be lost in the SWWebImage caching process.
```
```suggestion
    // the image-data and this causes this "meta" data to be lost in the SDWebImage caching process.
```

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
+        image = [image resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];
+      }
+    }
+    self.image = image;
```
Do we want to reset the `image` property even if it isn't changed when `resizeMode == repeat`? In general `sd_setImageWithURL` takes care of `self.image = image` and I may be afraid (having your comment in mind from the other PR) that setting `image` property may cause some extra operations on this `UIImageView`.

EDIT: Ok, I now saw the `SDWebImageAvoidAutoSetImage`. Do you think it would be worth to comment about it here? 😅

IjzerenHein replied:
This was the only way to perform the "post" processing correctly.
I feel that `SDWebImageAvoidAutoSetImage` is pretty much self explanatory

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
+  
+  // Repeat resize mode is handled by the UIImage. Use scale to fill
+  // so the repeated image fills the UIImageView.
+  self.contentMode = resizeMode == RCTResizeModeRepeat
```
Maybe we should move this change to the completion block of the `sd_setImageWithURL`?

IjzerenHein replied:
No that wouldn't work. The image post-process function only needs to be called whenever the repeat-mode changes (or when the cap-insets change, but that will be added in a later PR)

sjchmiela replied:
True, let me rephrase my concern:

If we set the `contentMode` immediately when the prop is set, on a slow device in the following scenario we would see a glitch:

- an image `X` is downloaded and presented with `resizeMode = A`
- props change: image changes to `Y`, `resizeMode` changes to `B`
  - `setSource:` is called, source updates, `needsReload` is set to `true`
  - `setResizeMode:` is called, `needsReload` is set to `true`, **`contentMode` is changed immediately**
  - `didSetProps:` is called, a new image is downloaded and presented

For a brief moment image `X` is presented with `contentMode` `B`.

IjzerenHein replied:
I understand your concern here and your point is entirely valid.

However, the issue of changing props and then consistently updating the view in one pass (synchronously) would still remain for other props like borders, shadows, etc.. It would just be solved for this particular case..

If you would want to solve this right, you would have to sync all the prop updates on the view (e.g. after the image load has completed). However, this would also affect animatable props and probably make them much update much slower. Personally, I don't think this is a route we would want to go down into.

sjchmiela replied:
Sorry for not making myself clear, what I wanted to propose was to _solve things right_ (as you named it) and sync all the prop updates. 😅 When it comes to animatable properties I don't see why would the implementation be slower for them (the only difference that I see would be that in `didSetProps` we would have to compare more values: previous vs next border radii, shadow, etc, which shouldn't be that much slower 🤔).

Looking into other image components (`RCTImageView` and `FFImageView`) I can see they both implement `setResizeMode:` the same way (set the value immediately), so maybe we could get away with it… but I certainly wish we could make sure the scenario above wouldn't happen.

Would you care to add a comment like `// TODO (@sjchmiela): Research props syncing <link to discussion or issue #7440>` so it reminds me to get back to this issue when we're done with major features?

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
 {
-  NSURL *imageURL = [RCTConvert NSURL:source[sourceUriKey]];
+  _source = source;
+  _needsReload = YES;
```
```suggestion
  // TODO: Implement equatable image source abstraction
  _needsReload = YES;
```
? 😃

IjzerenHein replied:
Totally :) I felt this was out-of-scope for this PR, that's why I didn't :)

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -4,9 +4,28 @@
+  if (self = [super init]) {
+    _needsReload = NO;
+    _resizeMode = RCTResizeModeCover;
+    self.contentMode = (UIViewContentMode)_resizeMode;
```
This may be just me, but I'm not particularly fond of just force casting that way. Do you think extracting this to a helper function `contentModeFromResizeMode` would be an overkill?

IjzerenHein replied:
I copied this from RN, but I do agree a helper function would be better. Will do 👍

sjchmiela replied:
Maybe we could:
- either use a static value (as we do with `RCTResizeModeCover`) and put `self.contentMode = UIViewContentModeScaleAspectFill` right after the `_resizeMode` (and maybe write a comment that these two should be kept in sync)
- or use `[self setResizeMode:RCTResizeModeCover]` so that we don't have an initializing code that is somehow special?

IjzerenHein replied:
I've added a helper function which does the conversion. In fact I've added a `EXImageTypes` class which now contains the other enums and helper functions as well, so that the EXImageView class stays a bit cleaner 👍

sjchmiela commented on packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageTypes.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class ExpoImageTypes {
```
When implementing `expo-notifications` I had to add several enum types to the package and I came up with this _enum pattern_, where for every JS-native enum we define an enum + a couple of helper functions. This has some advantages:
- an instance of an enum holds information about both JS and native value (coherency)
- putting different enums into different classes doesn't add too much bloat to one file
- with colocated functions we can convert back and forth between values of different realms
- you can use anything as native type (Strings too)
- an extra layer of abstraction between JS enum and native values (not exporting `ScaleType.FIT_CENTER` directly to JS)

Disadvantages of current implementations is `o(n)` searching when converting (could be mitigated with a simple hashmap creation).

I haven't tried to push to make this a convention in this repository, but so far I like it. What do you think about this pattern? An example would be:

https://github.com/expo/expo/blob/d09fc44c62ae96e0347e1909a774b1b93d866b49/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/enums/AudioContentType.java

IjzerenHein replied:
Hey! I like this approach and the effort in standardizing this kind of thing.  I've added a dedicated `ExpoImageResizeMode` enum, please let me know whether this is what you had in mind.

While making the change, I came across this bug:
- There seems to be an error in the AudioContentType example that you provided. This line should use `getNativeValue`: https://github.com/expo/expo/blob/d09fc44c62ae96e0347e1909a774b1b93d866b49/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/enums/AudioContentType.java#L39

sjchmiela commented on packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java:
```diff
@@ -40,9 +44,34 @@ public void setSource(ImageView view, @Nullable ReadableMap sourceMap) {
+    } else {
+      scaleType = ExpoImageTypes.getScaleType(resizeMode);
+    }
+    view.setScaleType(scaleType);
```
This implementation differs from the Android one when it comes to when the prop changes are applied. We could probably let it go, if this PR lands before the events PR I will simply rebase and split these changes.

IjzerenHein replied:
That is correct. There will be a lot more differences between the iOS/Android implementations at the view/drawing level

sjchmiela commented on packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java:
```diff
@@ -40,9 +44,34 @@ public void setSource(ImageView view, @Nullable ReadableMap sourceMap) {
 
+    RequestOptions options = new RequestOptions();
+
+    // Override the size for local assets. This ensures that
```
How (does?) does it work for remote assets? 😕

IjzerenHein replied:
It doesn't. Local assets have a so-called scale-suffix in the file-name (`@3x`) making it possible to who that image by default in it's "real" resolution. This is not possible for remotely loaded assets. For those kinds of images you will have to specify the `width` and `height` in the style or in the `source`.

sjchmiela reviewed (commented):
Thanks so much for this work! The results are looking great! 👍

IjzerenHein commented on packages/expo-image/src/Image.tsx:
```diff
@@ -11,10 +18,16 @@ export interface ImageProps extends AccessibilityProps {
+    const { style, resizeMode: resizeModeProp, ...restProps } = this.props;
+
+    const { resizeMode: resizeModeStyle, ...restStyle } = StyleSheet.flatten([style]) || {};
+    const resizeMode = resizeModeProp || resizeModeStyle || 'cover';
```
Will do!

IjzerenHein commented on packages/expo-image/src/ExpoImage.tsx:
```diff
@@ -1,12 +1,18 @@
+  let resolvedStyle = style;
 
-  return <NativeExpoImage {...props} source={resolvedSource} />;
+  if (!Array.isArray(resolvedSource)) {
```
Yes I believe so. The `source` as defined in the RN Image component can be an array of `ImageURISource`. These can be multiple sources and the native code can then select the best option for the given resolution.
https://reactnative.dev/docs/image/#source
https://github.com/facebook/react-native/blob/master/Libraries/Image/ImageSource.js#L85

The typeof of such an array is `object` and `resolveAssetSource` will just return that if that is the case.

IjzerenHein commented on packages/expo-image/src/ExpoImage.tsx:
```diff
@@ -1,12 +1,18 @@
-  return <NativeExpoImage {...props} source={resolvedSource} />;
+  if (!Array.isArray(resolvedSource)) {
+    const { width, height } = resolvedSource;
+    resolvedStyle = StyleSheet.flatten([{ width, height }, style]);
```
This code was taken primarily from the RN source:
https://github.com/facebook/react-native/blob/master/Libraries/Image/Image.ios.js#L103

I believe that it is used for Yoga layout purposes only and it only applies to `require(..)` assets. These assets contain the width/height and form the default when laying out the image component.

sjchmiela replied:
Would there be any difference if we didn't provide the `width` and `height`? I'm afraid polyfilling this makes a difference in how different images types are handled and rendered which is something I think we want to avoid. 🤔

IjzerenHein replied:
I will take a closer look! 👓

IjzerenHein replied:
Okay. So it turns out my suspicion was right here. The `{width, height}` styles are passed down to the Yoga layout system so it can use them as the `intrinsic size` of the image. This means that when you omit any size styles, then the View will use the size of the image itself. 

I've takes a look at doing this at the native level, and it can be done. However, because of the way RN implements this, it has to schedule a block on the UI-manager (which may cause it to wait), and it always causes the layout to be invalidated, causing unnecessary layout passes.
Doing this in JS is really the best performing way as the size information is passed to Yoga directly. I've added comments to the code elaborating on this.

sjchmiela replied:
I see… my personal opinion on that would be to treat all images alike, so either no user-set size = no image (intrinsic size `(0, 0)`) or to always use image's size as intrinsic size; I think this would be an improvement over current RN's API.

Let's discuss it (its feasibility) in the API proposal (soon to come)! 😃 

Until we decide otherwise I agree with you on the `{width, height}` passed to Yoga. 👍

IjzerenHein replied:
Let's do that 👍

IjzerenHein committed:
[expo-image] Add `resizeMode` feature

[expo-image] Add resizeMode for iOS

[expo-image] Add resize-mode for Android

resizeMode “repeat” is not yet supported. That feature requires more effort on Android and it is used very sparsly.

[expo-image] Add `resizeMode` prop to JS

[expo-image] Add resizeMode for iOS

[expo-image] Add resize-mode for Android

resizeMode “repeat” is not yet supported. That feature requires more effort on Android and it is used very sparsly.

[expo-image] JS improvements

[expo-image] iOS improvements after review

[expo-image] Add dedicated ExpoImageResizeMode enum on Android

[expo-image] Better resizeMode prop handling Android

[expo-image] Add test with no specified size

[expo-image] Add explanation about passing source width/height to yoga

[expo-image] Add intrinsic content size handling for remote assets (iOS)

[expo-image] Add .ico intrinsic content size test

# Conflicts:
#	packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java
#	packages/expo-image/ios/EXImageView.h
#	packages/expo-image/ios/EXImageView.m
#	packages/expo-image/src/Image.tsx

IjzerenHein committed:
[expo-image] Update build files

IjzerenHein committed:
[expo-image] Fixes after rebase

IjzerenHein committed:
[expo-image][android] Moved ResizeMode to enums folder & rename

IjzerenHein commented:
@sjchmiela I've rebased on the latest master/events. Let me know whether there are any blockers 👍

IjzerenHein committed:
[expo-image] Added explicit size test

IjzerenHein commented:
@sjchmiela Mind if I merge ?

sjchmiela commented on packages/expo-image/android/src/main/java/expo/modules/image/enums/ImageResizeMode.java:
```diff
@@ -0,0 +1,37 @@
+      }
+      return ImageResizeMode.UNKNOWN;
+    }
+  }
```
This file is unformatted (or formatted wrongly), it has a 4-space indentation (instead of 2-space) and doesn't have a newline at the end of file. I think the AS settings may be git-tracked only for root Android project. 😕

Would you mind reformatting this file? 🙂

IjzerenHein replied:
Whoops 😅 Reformatted it, thanks! 👍

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -2,16 +2,39 @@
 
 @implementation EXImageView
+{
+  __weak RCTBridge *_bridge;
```
Is there any reason for having two types of property declarations (one being with `@property`, other one with `{ properties… }`)? Does this one have any advantage over the other?

If not let's make the property declaration consistent with others:

```objc
@property (nonatomic, weak) RCTBridge *bridge;
```

IjzerenHein replied:
Yeah you're right that would be better. I've updated the code, thanks!

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -43,9 +85,33 @@ - (void)updateImage
 
-  [self sd_setImageWithURL:_imageURL
+  NSURL *imageURL = _source ? [RCTConvert NSURL:_source[sourceUriKey]] : nil;
+  NSNumber *scale = _source && _source[sourceScaleKey] ? _source[sourceScaleKey] : nil;
```
```suggestion
  NSNumber *scale = _source && _source[sourceScaleKey] ?: nil;
```
Could be written shorter with the Elvis operator

IjzerenHein replied:
Actually, your suggestion doesn't compile. I was able to write it shorter into

`NSNumber *scale = _source ? _source[sourceScaleKey] : nil;`

sjchmiela replied:
Ah, ok, sorry, didn't notice that the fallback value is `nil`. If so, I think we could just use

```objc
NSNumber *scale = _source[sourceScaleKey];
```

I ensured this should be ok by running

```objc
  NSDictionary *dict = nil;
  NSLog(@"%@", dict[@"abc"]);
```
which didn't crash the application (logged `(null)`).

IjzerenHein replied:
Works like a charm. Must be how Objective-C sends messages to objects that makes this work probably.

Updated the code, looks a lot slimmer 👍

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -79,6 +149,24 @@ - (SDExternalCompletionBlock)completionBlock
+    // When no explicit source image size was specified, use the dimensions
+    // of the loaded image as the intrinsic content size.
+    if (!width && !height) {
+      [self updateIntrinsicContentSize:image.size internalAsset:NO];
```
```suggestion
      [strongSelf updateIntrinsicContentSize:image.size internalAsset:NO];
```

IjzerenHein replied:
Whoops, I missed that one. Thanks, fixed it! 👍

sjchmiela commented on packages/expo-image/ios/EXImageView.m:
```diff
@@ -23,12 +46,31 @@ - (void)dealloc
 - (void)didSetProps:(NSArray<NSString *> *)changedProps
 {
-  if ([changedProps containsObject:@"source"]) {
+  if (_needsReload) {
```
We never set this to `NO` (apart from the initializer).

IjzerenHein replied:
Yup I noticed that as well. Fixed now!

sjchmiela commented on packages/expo-image/src/Image.tsx:
```diff
@@ -1,21 +1,26 @@
   // function, so we can't use it there. So we pass the unresolved source
   // to "native components" and they decide whether to resolve the value
   // or not.
+  resizeMode?: ImageResizeMode;
```
Could you please put it not immediately after the comment describing the `source` prop (so either before the comment or after the `source` prop) so the comment stays colocated with the prop it talks about?

IjzerenHein replied:
Ah I hadn't noticed that. I thought the props were in alphabetical order that's why I put it there.
I've moved it to the bottom for now.
Would you reckon having the props in alphabetical order is a good thing?

sjchmiela reviewed (approved):
Apart from these couple of comments it looks good! 👍 👍 👍

IjzerenHein committed:
[expo-image][ios] Add class for types & convert functions

IjzerenHein committed:
[expo-image][ios] Clarify resize-mode code

IjzerenHein committed:
[expo-image][ios] Fixes after review

IjzerenHein committed:
[expo-image][ios] Update pod files

IjzerenHein committed:
[expo-image][android] Fix code style

IjzerenHein committed:
[expo-image] Fix prop location

IjzerenHein committed:
[expo-image] Update build files

IjzerenHein committed:
[expo-image][ios] Simply dictionary access
